src/guiguts/utilities.py

```diff
@@ -360,3 +360,318 @@ def remove_diacritics(cls, string: str) -> str:
+
+            wordcount = len(words)
+            # Add paragraph indentation to first word if there is one.
+            if wordcount > 0:
+                words[0] = f"{indent1}{words[0]}"
+            word_len = [0] * wordcount  # Length of each word (excluding spaces).
+            space_len = [0] * wordcount  # Length of the space after after this word.
+            for j in range(wordcount):
+                if re.findall(r" $", words[j]):
+                    word_len[j] = len(words[j]) - 1
+                    space_len[j] = 2
+                else:
+                    word_len[j] = len(words[j])
+                    space_len[j] = 1
+            # First word already has 'indent1' added and will not be indented further.
+            if wordcount > 0:
+                word_len[0] -= len(indent2)
+
+        def compute_output() -> None:
+            """Compute global 'output' list from 'wordcount', 'words' list,
+            lastbreak and linkbreak"""
+
+            nonlocal output, words, wordcount, lastbreak, linkbreak
+
+            output = []
+            terminus = wordcount - 1
+            while terminus >= 0:
+                # NB The lower bound is inclusive, the upper bound is noninclusive.
+                my_str = " ".join(words[lastbreak + 1 : terminus + 1]) + "\n"
+                output.append(my_str)
+                terminus = lastbreak
+                lastbreak = linkbreak[lastbreak]
+            output.reverse()
+            # Add the indent to all but the first line.
+            mapped = map(lambda s: f"{indent2}{s}", output[1 : len(output)])
+            output = [output[0]]
+            for m in mapped:
+                output.append(m)
+
+        def reflow_para() -> None:
+            """Rewrap pragraph"""
+
+            nonlocal words
+            nonlocal lastbreak, linkbreak, wordcount, output
+
+            if len(words) == 0:
+                return
+            reflow_penalties()
+            lastbreak = 0
+            lastbreak = reflow_trial()
+            compute_output()
+            # Restore spaces in ellipses
+            tmp = []
+            for line in output:
+                tmp.append(re.sub(r"\x9f", " ", line))
+            output = tmp.copy()
+            print_lines(output)
+            words = []
+
+        def process(line: str) -> None:
+            """Process a line by appending each word to global 'words' list.
+            If the line is blank, reflow the paragraph of words in 'words'.
+            """
+            nonlocal words
+
+            # Protect " . . ." ellipses by replacing space with unused byte \x9f.
+            line = re.sub(r" \. \. \.", r"\x9f.\x9f.\x9f.", line)
+            line = re.sub(r"\. \. \.", r".\x9f.\x9f.", line)
+            # Splitting assumes a single space between words in 'line'
+            line = re.sub(r"\s+", " ", line)
+            # Remove any extra space at end of line.
+            line = re.sub(r"\s$", "", line)
```

Comment:
Should that be 

line = re.sub(r"\s`+`$", "", line)

In case there are multiple spaces at the end of a line?

Reply:
Thanks for checking so carefully. In this case, it's not needed, because the line before replaces any multiple spaces with a single space.

Reply:
> Thanks for checking so carefully. In this case, it's not needed, because the line before replaces any multiple spaces with a single space.

Oops, I saw that, but didn't bring it forward 🤦‍♀️

Reply:
Well, I actually added the `+` in the code 🤦‍♀️, but once the `\s+` was in the code, my brain spotted the exact same string on the line above. :)